### PR TITLE
Docs

### LINKED ISSUE #87: Add documentation
- Blurb
- Quick Start
- Tutorial
- API

### PR DESCRIPTION
**Pull request by jamesopstad**
resolves #87

This adds documentation to the readme. I've tried to keep the tutorial as succinct as possible but think it's a nice way to introduce the key concepts.

### CODE REVIEW

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (jamesopstad) [inline packages/vite-plugin-cloudflare/src/plugin-config.ts:1]: Changes in this file are just to make the naming more consistent with the docs.

File: README.md [hunk 1]
```diff
@@ -1,10 +1,376 @@
...
+Your Worker code runs inside [workerd](https://github.com/cloudflare/workerd), matching the production behaviour as closely as possible and providing confidence as you develop and deploy your applications.
 
-```bash
-npm install https://pkg.pr.new/flarelabs-net/vite-plugin-cloudflare/@flarelabs-net/vite-plugin-cloudflare@main --save-dev
+### Features
+
+- Provides direct access to Workers runtime APIs and bindings
+- Supports Workers Assets, enabling you to build static sites, SPAs and full-stack applications
+- Leverages Vite's hot-module reloading for consistently fast updates
+- Supports `vite preview` for previewing your build output in the Workers runtime prior to deployment
```
Comment (jamesopstad) [inline README.md:13]: We could add some emojis here. Not my area of expertise so suggestions welcome!

File: README.md [hunk 2]
```diff
@@ -1,10 +1,376 @@
...
+
+> [!NOTE]
+> When using the Cloudflare Vite plugin, the Worker config (e.g. `wrangler.toml`) that you provide is the input configuration file.
+> A separate output `wrangler.json` file is created when you run `vite build`.
+> This output file is a snapshot of your configuration at the time of the build and is the configuration used for preview and deployment.
+
+### Configure not found handling
+
+If you run `vite dev`, you will see that your app is running.
+If you navigate to a different path, however, you will get a 404 response.
```
Comment (dario-piotrowicz) [inline README.md:119]:
```suggestion
If you navigate to a path different to `/`, however, you will get a 404 response.
```
or
```suggestion
If you navigate to a path other than the index one, however, you will get a 404 response.
```

File: README.md [hunk 3]
```diff
@@ -1,10 +1,376 @@
...
+This is because the not found handling is now being handled by the Cloudflare plugin rather than Vite and replicates the production behaviour that is specified in your configuration.
+To make all not found requests direct to the `index.html` file, set the value to `single-page-application` in your Worker config.
+
+```toml
+# wrangler.toml
+
+assets = { not_found_handling = "single-page-application" }
+```
+
+Now, if you navigate to a different path, you will still see your application.
```
Comment (dario-piotrowicz) [inline README.md:129]: same as above, since we haven't specified what path the user navigates to, different path is a bit ambiguous in my opinion

File: README.md [hunk 4]
```diff
@@ -1,10 +1,376 @@
...
+        name: 'Cloudflare',
+      });
+    }
+
+    return env.ASSETS.fetch(request);
+  },
+} satisfies ExportedHandler<Env>;
+```
+
+The Worker above will be invoked for any not found path.
```
Comment (IgorMinar) [inline README.md:203]:
```suggestion
The Worker above will be invoked for any request not matching a static asset.
```

File: README.md [hunk 5]
```diff
@@ -1,10 +1,376 @@
...
+It returns a JSON response if the `pathname` starts with `/api/` and otherwise passes the incoming request through to the asset binding.
+This means that for paths that do not start with `/api/`, the `not_found_handling` behaviour defined in the Worker config will be evaluated and the `index.html` file will be returned.
```
Comment (IgorMinar) [inline README.md:205]:
```suggestion
This means that for paths that do not start with `/api/`, the `not_found_handling` behaviour defined in the Worker config will be evaluated and the `index.html` file will be returned to enable single page application navigations.
```

File: README.md [hunk 6]
```diff
@@ -1,10 +1,376 @@
...
+    </>
+  );
+}
+
+export default App;
 ```
+
+Now, if you click the button, it will display 'Name from API is: Cloudflare'.
+
+Try incrementing the counter and then changing the `name` that is returned in `api/index.ts`.
```
Comment (IgorMinar) [inline README.md:272]: consider being a bit more explicit here, as the instructions are a bit vague and the impact might not be understood, for example (feel free to rephrase):

> Let's now see hot module reloading in action. Try incrementing the counter to initialize application state in the browser. Then edit `api/index.ts` by changing the `name` it returns to `"Cloudflare Workers"`. If you now click the button again, it will display the new `name` while preserving the previously set counter value! You can now iterate on your app quickly without losing any of the UI state in between edits.

File: README.md [hunk 7]
```diff
@@ -1,10 +1,376 @@
...
+
+Run `vite build` to build your application.
+
+If you inspect the `dist` directory, you will see that it contains two subdirectories: `client` and `api`.
+The `api` directory contains your Worker code and the output `wrangler.json` configuration.
+
+### Preview your application
+
+Run `vite preview` to validate that your application runs as expected.
+This command will run your build output locally in the Workers runtime.
```
Comment (IgorMinar) [inline README.md:285]:
```suggestion
This command will run your production build output locally in the Workers runtime, without HMR or other development support code.
```

File: README.md [hunk 8]
```diff
@@ -1,10 +1,376 @@
...
+
+### Deploy to Cloudflare
+
+Run `wrangler deploy` to deploy your application to Cloudflare.
+This command will automatically use the output `wrangler.json` that was included in the build output.
+
+### Next steps
+
+In this tutorial, we created an SPA that could be deployed using Workers Assets.
```
Comment (IgorMinar) [inline README.md:294]:
```suggestion
In this tutorial, we created an SPA that could be deployed using a Worker with Worker Assets.
```

File: README.md [hunk 9]
```diff
@@ -1,10 +1,376 @@
...
+
+### `cloudflare`
+
+The `cloudflare` plugin should be included in the Vite `plugins` array:
+
+```ts
+// vite.config.ts
+
+import { defineConfig } from 'vite';
+import { cloudflare } from '@cloudflare/vite-plugin';
```
Comment (IgorMinar) [inline README.md:312]: are we moving away from the default import?

File: README.md [hunk 10]
```diff
@@ -1,10 +1,376 @@
...
+
+- `viteEnvironment?: { name?: string }`
+
+  Optional Vite environment options.
+  By default, the environment name is the Worker name with `-` characters replaced with `_`.
+  Setting the name here will override this.
+
+## Migrating from `wrangler dev`
+
+Migrating from `wrangler dev` should be straightforward but there are a few key differences to highlight:
```
[the remark this answers is not in the document]
Comment (IgorMinar) [reply README.md:365]: maybe just a reference to install the plugin as mentioned in the quick start above?

File: packages/vite-plugin-cloudflare/src/plugin-config.ts [hunk 11]
```diff
@@ -7,13 +7,20 @@ import type { Unstable_Config } from 'wrangler';
...
-interface PluginWorkerConfig {
-	configPath: string;
+interface BaseWorkerConfig {
 	viteEnvironment?: { name?: string };
 }
 
-export interface PluginConfig extends Partial<PluginWorkerConfig> {
-	auxiliaryWorkers?: PluginWorkerConfig[];
+interface EntryWorkerConfig extends BaseWorkerConfig {
+	configPath?: string;
```
Comment (IgorMinar) [inline packages/vite-plugin-cloudflare/src/plugin-config.ts:15]: could you also add light jsdocs please?

File: README.md [hunk 12]
```diff
@@ -1,10 +1,376 @@
...
+// src/index.ts
+
+export default {
+  fetch() {
+    return new Response(`Running in ${navigator.userAgent}!`);
+  },
+};
+```
+
+## Tutorial
```
[the remark this answers is not in the document]
Comment (IgorMinar) [reply README.md:64]: it's not that long, so suggest we keep it inline for now.

**Review by IgorMinar — APPROVED** [review]
Overall this looks really good and comprehensive to me. I left a few notes here and there with clarifications, but I think overall this make sense to me.

I'm approving this assuming that we resolve all the comments via edits an agreement to take no action.

Thank you @jamesopstad!

File: README.md [hunk 13]
```diff
@@ -1,10 +1,376 @@
...
+
+export default defineConfig({
+  plugins: [react(), cloudflare()],
+});
+```
+
+### Create your Worker config file
+
+```toml
+# wrangler.toml
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply README.md:103]: I was expecting this too.

File: README.md [hunk 14]
```diff
@@ -1,10 +1,376 @@
...
+
+assets = { not_found_handling = "single-page-application" }
+```
+
+Now, if you navigate to a different path, you will still see your application.
+
+For a purely front-end application, you could now proceed to build and deploy your application.
```
Comment (petebacondarwin) [inline README.md:131]: Perhaps put in a heading that is something like "Adding an API Worker" or similar.
The next heading being "Add the `@cloudflare/workers-types` dependency and configure TypeScript" felt jarring.

File: README.md [hunk 15]
```diff
@@ -1,10 +1,376 @@
...
+```
+
+```jsonc
+// tsconfig.json
+
+{
+  "files": [],
+  "references": [
+    { "path": "./tsconfig.app.json" },
+    { "path": "./tsconfig.node.json" },
```
Comment (petebacondarwin) [inline README.md:160]: Does the Vite React creation tool generate the `tsconfig.app.json` and `tsconfig.node.json`?
If so, how come it doesn't already include the TypeScript dependency?

**Review by petebacondarwin — APPROVED** [review]
LGTM once Igor and Dario's comments have been addressed

**Comment by dario-piotrowicz** [comment]
A minor comment that just occurred to me, given the recent push for `wrangler.json` (https://github.com/cloudflare/workers-sdk/pull/7676, etc...) should we maybe switch our wrangler files into jsons? and also refer to the json format in the docs?

I think that'd make sense, it also doesn't need to be addressed in this PR (since it would probably be good to also update the config files in the various playground apps)

One concern I have docs wise is that now we get a bit of a stronger distinction between the userland/input `wrangler.toml` and the output `wrangler.json`, if we had `wrangler.json` both as the input and output that could potentially confuse some people if not conveyed very clearly 🤔

File: README.md (hunk 15 above)
Comment (jamesopstad) [reply README.md:155]: It does generate those files and includes the TypeScript dependency. Not quite sure what you mean?

File: README.md [hunk 16]
```diff
@@ -1,10 +1,376 @@
...
+
+  Optional Vite environment options.
+  By default, the environment name is the Worker name with `-` characters replaced with `_`.
+  Setting the name here will override this.
+
+- `persistState?: boolean | { path: string }`
+
+  An optional override for state persistence.
+  By default, state is persisted to `.wrangler/state/v3`.
+  A custom `path` can be provided or, alternatively, persistence can be disabled by setting the value to `false`.
```
[the remark this answers is not in the document]
Comment (jamesopstad) [reply README.md:337]: I've separated out the `v3` subdirectory so this is clearer.

File: README.md [hunk 17]
```diff
@@ -1,10 +1,376 @@
...
+
+```ts
+// vite.config.ts
+
+import { defineConfig } from 'vite';
+import react from '@vitejs/plugin-react';
+import { cloudflare } from '@cloudflare/vite-plugin';
+
+export default defineConfig({
+  plugins: [react(), cloudflare()],
```
[the remark this answers is not in the document]
Comment (jamesopstad) [reply README.md:94]: See above comment.

File: README.md (hunk 9 above)
Comment (jamesopstad) [reply README.md]: See other comment about possibly importing a different package. As far as I remember, we never used a default export.

File: README.md [hunk 18]
```diff
@@ -1,10 +1,376 @@
...
 
-```bash
-npm install https://pkg.pr.new/flarelabs-net/vite-plugin-cloudflare/@flarelabs-net/vite-plugin-cloudflare@main --save-dev
+### Features
+
+- Provides direct access to Workers runtime APIs and bindings
+- Supports Workers Assets, enabling you to build static sites, SPAs and full-stack applications
+- Leverages Vite's hot-module reloading for consistently fast updates
+- Supports `vite preview` for previewing your build output in the Workers runtime prior to deployment
+
```
[the remark this answers is not in the document]
Comment (jamesopstad) [reply README.md:16]: It might be a bit wordy to include this here before the Quick Start. Is it not clear that it's included as a dependency to install?

File: README.md (hunk 14 above)
Comment (jamesopstad) [reply README.md]: I've changed the heading to 'Configure TypeScript for your Worker code'. Is that OK? Alternatively I could split the tutorial with two sub-headings and knock all the other headings down a level.

File: README.md [hunk 19]
```diff
@@ -1,10 +1,382 @@
...
+```toml
+# wrangler.toml
+
+name = "cloudflare-vite-tutorial"
+compatibility_date = "2024-12-30"
+assets = { not_found_handling = "single-page-application" }
+```
+
+We have set the [`not_found_handling`](https://developers.cloudflare.com/workers/static-assets/routing/#not_found_handling--404-page--single-page-application--none) value to `single-page-application`.
```
Comment (ToriLindsay) [inline README.md:110]:
```suggestion
You have set the [`not_found_handling`](https://developers.cloudflare.com/workers/static-assets/routing/#not_found_handling--404-page--single-page-application--none) value to `single-page-application`.
```
Or "has been set to", if you prefer

File: README.md [hunk 20]
```diff
@@ -1,10 +1,382 @@
...
+
+In this tutorial, we're going to create a React SPA that can be deployed as a Worker with Workers Assets.
+We'll then add an API Worker that can be accessed from the front-end code.
+We will develop, build and preview the application using Vite before finally deploying to Cloudflare.
+
+### Set up and configure the React SPA
+
+#### Scaffold a Vite project
+
+Let's start by creating a React TypeScript project with Vite.
```
Comment (ToriLindsay) [inline README.md:72]:
```suggestion
Start by creating a React TypeScript project with Vite.
```

File: README.md [hunk 21]
```diff
@@ -1,10 +1,382 @@
...
+    </>
+  )
+}
+
+export default App
+```
+
+Now, if you click the button, it will display 'Name from API is: Cloudflare'.
+
+Let's have a look at hot module reloading in action.
```
Comment (ToriLindsay) [inline README.md:272]:
```suggestion
```

File: README.md [hunk 22]
```diff
@@ -1,10 +1,382 @@
...
+Increment the counter to update the application state in the browser.
```
Comment (ToriLindsay) [inline README.md:273]:
```suggestion
You can see hot module reloading in action by Incrementing the counter to update the application state in the browser.
```

File: README.md [hunk 23]
```diff
@@ -1,10 +1,382 @@
...
+```toml
+# wrangler.toml
+
+name = "cloudflare-vite-tutorial"
+compatibility_date = "2024-12-30"
+assets = { not_found_handling = "single-page-application", binding = "ASSETS" }
+main = "./api/index.ts"
 ```
+
+The assets `binding` defined here will allow us to access the assets functionality from our Worker.
```
Comment (ToriLindsay) [inline README.md:176]:
```suggestion
The assets `binding` defined here will allow you to access the assets functionality from your Worker.
```

File: README.md [hunk 24]
```diff
@@ -1,10 +1,382 @@
...
+We have set the [`not_found_handling`](https://developers.cloudflare.com/workers/static-assets/routing/#not_found_handling--404-page--single-page-application--none) value to `single-page-application`.
+This means that all not found requests will serve the `index.html` file.
+With the Cloudflare plugin, the `assets` routing configuration is used in place of Vite's default behaviour.
+This ensures that your application's routing works the same way while developing and is it does when deployed to production.
+
+Note that the [`directory`](https://developers.cloudflare.com/workers/static-assets/binding/#directory) field is not used when configuring assets with Vite.
+The `directory` in the output configuration will automatically point to the client build output.
+
+> [!NOTE]
+> When using the Cloudflare Vite plugin, the Worker config (e.g. `wrangler.toml`) that you provide is the input configuration file.
```
Comment (ToriLindsay) [inline README.md:119]:
```suggestion
> When using the Cloudflare Vite plugin, the Worker config (for example, `wrangler.toml`) that you provide is the input configuration file.
```
Unless you meant "i.e."? In that case, just change it to simply "(wrangler.toml)"

File: README.md [hunk 25]
```diff
@@ -1,10 +1,382 @@
...
+```
+
+The Worker above will be invoked for any request not matching a static asset.
+It returns a JSON response if the `pathname` starts with `/api/` and otherwise passes the incoming request through to the asset binding.
+This means that for paths that do not start with `/api/`, the `not_found_handling` behaviour defined in the Worker config will be evaluated and the `index.html` file will be returned, enabling SPA navigations.
+
+#### Call the API from the client
+
+Edit `src/App.tsx` so that it includes an additional button that calls the API and sets some state.
+You can replace the file contents with the following code.
```
Comment (ToriLindsay) [inline README.md:209]:
```suggestion
Replace the file contents with the following code:
```

File: README.md [hunk 26]
```diff
@@ -1,10 +1,382 @@
...
+
+Run `vite preview` to validate that your application runs as expected.
+This command will run your build output locally in the Workers runtime, closely matching its behaviour in production.
+
+#### Deploy to Cloudflare
+
+Run `npm exec wrangler deploy` to deploy your application to Cloudflare.
+This command will automatically use the output `wrangler.json` that was included in the build output.
+
+### Next steps
```
Comment (ToriLindsay) [inline README.md:295]: Maybe change to something more specific like "After deployment" to make more scannable?

File: README.md [hunk 27]
```diff
@@ -1,10 +1,382 @@
...
+- `auxiliaryWorkers?: Array<AuxiliaryWorkerConfig>`
+
+  An optional array of auxiliary workers.
+  You can use [service bindings](https://developers.cloudflare.com/workers/runtime-apis/bindings/service-bindings/) to call auxiliary workers from your main (entry) Worker.
+  All requests are routed through your entry Worker.
+  During the build, each Worker is output to a separate subdirectory of `dist`.
+
+> [!NOTE]
+> When running `wrangler deploy`, only your main (entry) Worker will be deployed.
+> If using multiple Workers, it is your responsibility to deploy them individually.
```
Comment (ToriLindsay) [inline README.md:352]:
```suggestion
> If using multiple Workers, each must be deployed individually.
```
The "your responsibility" just sounded kind of intense to me.

File: README.md [hunk 28]
```diff
@@ -1,10 +1,382 @@
...
+
+- `viteEnvironment?: { name?: string }`
+
+  Optional Vite environment options.
+  By default, the environment name is the Worker name with `-` characters replaced with `_`.
+  Setting the name here will override this.
+
+## Migrating from `wrangler dev`
+
+Migrating from `wrangler dev` is straightforward and you can follow the instructions in the [Quick Start](#quick-start) to get started.
```
Comment (ToriLindsay) [inline README.md:369]:
```suggestion
Migrating from `wrangler dev` is a simple process and you can follow the instructions in the [Quick Start](#quick-start) to get started.
```
Sounded too casual to me

File: README.md [hunk 29]
```diff
@@ -1,10 +1,382 @@
...
+### Input and output Worker config files
+
+In the Vite integration, your Worker config file (e.g. `wrangler.toml`) is the input configuration and a separate output configuration is created as part of the build.
+This output file is a snapshot of your configuration at the time of the build and is modified to reference your build artifacts.
+It is the configuration that is used for preview and deployment.
+
+### Redundant fields in the Wrangler config file
+
+There are various options in the Worker config file that are ignored when using Vite, as they are either no longer applicable or are replaced by Vite equivalents.
+If these options are provided then warnings will be printed to the console with suggestions for how to proceed.
```
Comment (ToriLindsay) [inline README.md:381]:
```suggestion
If these options are provided, then warnings will be printed to the console with suggestions for how to proceed.
```

File: README.md [hunk 30]
```diff
@@ -1,10 +1,382 @@
...
+> A separate output `wrangler.json` file is created when you run `vite build`.
+> This output file is a snapshot of your configuration at the time of the build and is modified to reference your build artifacts.
+> It is the configuration that is used for preview and deployment.
+
+#### Run the development server
+
+Run `npm run dev` to verify that your application is working as expected.
+
+For a purely front-end application, you could now proceed to build (`npm run build`), preview (`npm run preview`) and deploy (`npm exec wrangler deploy`) your application.
+We're going to go a step further, however, and add an API Worker.
```
Comment (ToriLindsay) [inline README.md:129]:
```suggestion
However, this tutorial will show you how to go a step further and add an API Worker.
```

File: README.md [hunk 31]
```diff
@@ -1,10 +1,382 @@
...
+Your Worker code runs inside [workerd](https://github.com/cloudflare/workerd), matching the production behaviour as closely as possible and providing confidence as you develop and deploy your applications.
+
+### Features
+
+- Provides direct access to Workers runtime APIs and bindings
+- Supports Workers Assets, enabling you to build static sites, SPAs and full-stack applications
+- Leverages Vite's hot module replacement for consistently fast updates
+- Supports `vite preview` for previewing your build output in the Workers runtime prior to deployment
+
+## Quick Start
```
Comment (ToriLindsay) [inline README.md:17]:
```suggestion
## Quick start
```

**Review by ToriLindsay — APPROVED** [review]
Just left a bunch of grammar/clarity suggestions. Feel free to ignore any that don't make sense for the context.
Note: We always use [the oxford comma](https://developers.cloudflare.com/style-guide/grammar/punctuation-marks-and-symbols/commas/).
Also, I noticed that it was jumping between the first, second, and third person a lot, so I removed the "we", "us", "let's", etc. to keep a little more consistent.

File: packages/vite-plugin-cloudflare/src/index.ts [hunk 32]
```diff
@@ -26,6 +26,10 @@ import { getWarningForWorkersConfigs } from './workers-configs';
 import type { PluginConfig, ResolvedPluginConfig } from './plugin-config';
 import type { Unstable_RawConfig } from 'wrangler';
 
+/**
+ * Cloudflare Vite plugin.
+ * Accepts an optional {@link PluginConfig} object.
```
Comment (dario-piotrowicz) [inline packages/vite-plugin-cloudflare/src/index.ts:31]: Should we have a more detailed description?

For the parameter, why not a standard [`@param` tag](https://jsdoc.app/tags-param)? 😮

File: README.md [hunk 33]
```diff
@@ -1,10 +1,458 @@
...
+
+- `viteEnvironment?: { name?: string }`
+
+  Optional Vite environment options.
+  By default, the environment name is the Worker name with `-` characters replaced with `_`.
+  Setting the name here will override this.
+
+## Worker environments
+
+A Worker config file may contain configuration for multiple [environments](https://developers.cloudflare.com/workers/wrangler/environments/).
```
Comment (petebacondarwin) [inline README.md:368]:
```suggestion
A Worker config file may contain configuration for multiple [Wrangler environments](https://developers.cloudflare.com/workers/wrangler/environments/).
```

Comment (petebacondarwin) [reply README.md:368]: or possibly

```suggestion
A Worker config file may contain configuration for multiple [Cloudflare environments](https://developers.cloudflare.com/workers/wrangler/environments/).
```

File: README.md [hunk 34]
```diff
@@ -1,10 +1,458 @@
...
+vars = { MY_VAR = "Top-level var" }
+
+[env.staging]
+vars = { MY_VAR = "Staging var" }
+
+[env.production]
+vars = { MY_VAR = "Production var" }
+```
+
+If you run `CLOUDFLARE_ENV=production vite build` then the output `wrangler.json` file generated by the build will be a flattened configuration for the 'production' environment.
```
Comment (dario-piotrowicz) [inline README.md:388]: We does this only specify `vite build` without any mention of `vite dev`?

File: README.md [hunk 35]
```diff
@@ -1,10 +1,458 @@
...
+This combines [top-level only](https://developers.cloudflare.com/workers/wrangler/configuration/#top-level-only-keys), [inheritable](https://developers.cloudflare.com/workers/wrangler/configuration/#inheritable-keys), and [non-inheritable](https://developers.cloudflare.com/workers/wrangler/configuration/#non-inheritable-keys) keys.
+The value of `MY_VAR` will therefore be `'Production var'`.
+The name of the Worker will be `'my-worker-production'`
+This is because the environment name is automatically appended to the top-level Worker name.
```
Comment (dario-piotrowicz) [inline README.md:392]:
```suggestion
The name of the Worker will be `'my-worker-production'`, this is because the environment name is automatically appended to the top-level Worker name.
```

Comment (petebacondarwin) [inline README.md:392]: Is there a docs page we could link to that explains this? I think not but if there is...

File: README.md [hunk 36]
```diff
@@ -1,10 +1,458 @@
...
+
+> [!NOTE]
+> The default Vite environment name for a Worker is always the top-level name.
+> This enables you to reference the Worker consistently in your Vite config when using multiple environments.
+
+Running `vite dev` or `vite build` without providing `CLOUDFLARE_ENV` will use the default top-level environment.
+The value of `MY_VAR` will therefore be `'Top-level var'`.
+As Worker environments are applied at dev and build time, specifying `CLOUDFLARE_ENV` when running `vite preview` or `wrangler deploy` will have no effect.
```
Comment (petebacondarwin) [inline README.md:400]: Feels like this should be called out in a `!NOTE` block?

File: README.md (hunk 33 above)
Comment (petebacondarwin) [reply README.md:368]: Oh I see below that you use the term "Worker environment" - so perhaps that would be best here.

File: README.md [hunk 37]
```diff
@@ -1,10 +1,458 @@
...
+CLOUDFLARE_ENV=staging
+```
+
+```sh
+# .env.production
+
+CLOUDFLARE_ENV=production
 ```
+
+Now, if you run `vite build`, the 'production' mode will be used by default and the 'production' environment will be selected.
```
Comment (petebacondarwin) [inline README.md:438]: I think this could be more explicit, especially for people who don't understand modes and environments.

Perhaps along the lines of:

```suggestion
By default the `vite build`, uses the 'production' mode. Vite will therefore load the `.env.production` file to get environment variables to be used in the build. Since the `.env.production` file contains `CLOUDFLARE_ENV=production` the Cloudflare Vite plugin will select the 'production' Worker environment configuration.
```

**Review by petebacondarwin — COMMENTED** [review]
Added some thoughts about the environments section.

File: README.md (hunk 35 above)
Comment (jamesopstad) [reply README.md:392]: It's only mentioned in this 'Configuration' section - https://developers.cloudflare.com/workers/wrangler/environments/#configuration.

File: packages/vite-plugin-cloudflare/src/index.ts (hunk 32 above)
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/index.ts:31]: To be honest, I just followed Vite's example here - https://github.com/vitejs/vite/blob/4f5845a3182fc950eb9cd76d7161698383113b18/packages/vite/src/node/config.ts#L137-L141.

Feel free to make a suggestion.

File: README.md (hunk 34 above)
Comment (jamesopstad) [reply README.md:388]: I couldn't find a way of including both without it becoming overly long and confusing (there's no output `wrangler.json` in dev for example). I have called out either side of this paragraph that it can be applied at dev and build time. I can have another think.

File: README.md (hunk 35 above)
Comment (jamesopstad) [reply README.md:392]: Sorry, there was supposed to be a full stop. A comma doesn't work grammatically.

```suggestion
The name of the Worker will be `'my-worker-production'`.
This is because the environment name is automatically appended to the top-level Worker name.
```

File: README.md (hunk 36 above)
Comment (jamesopstad) [reply README.md:400]: There's already a `!NOTE` block directly above so it felt a bit weird having them next to each other. Any suggestions?

File: README.md (hunk 37 above)
Comment (jamesopstad) [reply README.md]: [e0d0a21](https://github.com/flarelabs-net/vite-plugin-cloudflare/pull/119/commits/e0d0a2124631758d957adc445e01201ed29832a8)

File: README.md (hunk 33 above)
Comment (jamesopstad) [reply README.md]: I've updated it to 'Worker environments'. Let me know if you would rather I use the term 'Cloudflare environments' throughout the readme though. I'm not sure which is better.

File: README.md (hunk 34 above)
Comment (dario-piotrowicz) [reply README.md:388]: yeah I can see that you did mention "dev or build time", maybe that's good enough? 🤔 

In any case I'm fine with the current paragraph, we can always improve it later if needed

Comment (jamesopstad) [reply README.md:388]: I've added an additional paragraph - [886873f](https://github.com/flarelabs-net/vite-plugin-cloudflare/pull/119/commits/886873f6fa7528620dc5cdc83d64ef689a18a5ae).

File: README.md (hunk 36 above)
Comment (jamesopstad) [reply README.md]: I've fixed this by adding an additional paragraph in between - [886873f](https://github.com/flarelabs-net/vite-plugin-cloudflare/pull/119/commits/886873f6fa7528620dc5cdc83d64ef689a18a5ae).

Let me know if that's OK.

File: packages/vite-plugin-cloudflare/src/index.ts (hunk 32 above)
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/index.ts:31]: Thanks. Let's add the JSDoc once the readme is in workers-sdk so it points to the right place.

File: README.md (hunk 35 above)
Comment (petebacondarwin) [reply README.md:392]: It's pretty minimal. Let's leave it for now
